sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/StartingPointShardsFinder.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.kinesis;
+
+import com.amazonaws.services.kinesis.model.Shard;
+import com.amazonaws.services.kinesis.model.ShardIteratorType;
+import com.google.common.collect.Sets;
+
+import java.io.Serializable;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is responsible for establishing the initial set of shards that existed at the given
+ * starting point.
+ */
+class StartingPointShardsFinder implements Serializable {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StartingPointShardsFinder.class);
+
+  /**
+   * <p>Finds all the shards at the given startingPoint. This method starts by gathering the oldest
+   * shards in the stream and considers them as initial shards set. Then it validates the shards
+   * by getting an iterator at the given starting point and trying to read some records. If shard
+   * passes the validation then it is added to the result shards set. If not then it is regarded
+   * as expired and its successors are taken into consideration. This step is repeated until all
+   * valid shards are found.</p>
```

Comment:
Some shards might get closed while this is happening. Will the code still work correctly in that case? If so (i.e. if the code doesn't care if they are valid), then why validate shards at all?

Reply:
Shards might get closed during this operation but this will affect only the case when LATEST marker was passed as the staringPoint so only when you want to read the stream from now on. I didn't analyse what could happen in such situation but in the worst case you could just start the pipeline again.
Maybe shards validity is not the best wording here. By saying that shard is valid at a given startingPoint, I really mean that it can be read from at that point. So every shard is valid for the staringPoints since it's creation to it's termination.

Reply:
Not sure I understand the point about "you could just start the pipeline again". Do you mean that the current code may produce a pipeline that won't work and the only fix will be restarting it? That seems concerning and I'd rather not merge until this is resolved.

Reply:
I have an idea for the fix, I'll submit it by Monday.

Reply:
Was this addressed?

Reply:
Yes. Initially I was thinking about repeating part of that operation until all starting shards are found, but it would not be pretty. Then I realised that this situation should actually be treated just like any other exceptional situation so I remedied the issue by throwing an exception when the problem is found (see line 140).